impl/src/main/java/org/pentaho/platform/genericfile/DefaultGenericFileService.java

```diff
@@ -319,8 +310,79 @@ public void restoreFiles( @NonNull List<GenericFilePath> paths ) throws Operatio
 
   @Override
   public void restoreFile( @NonNull GenericFilePath path ) throws OperationFailedException {
-    getOwnerFileProvider( path )
-      .orElseThrow( () -> new NotFoundException( String.format( "Path not found '%s'.", path ) ) )
-      .restoreFile( path );
+    getOwnerFileProvider( path ).restoreFile( path );
+  }
+
+  @Override
+  public boolean renameFile( @NonNull GenericFilePath path, @NonNull String newName ) throws OperationFailedException {
+    return getOwnerFileProvider( path ).renameFile( path, newName );
+  }
+
+  @Override
+  public void copyFiles( @NonNull List<GenericFilePath> paths, @NonNull GenericFilePath destinationPath )
+    throws OperationFailedException {
+    BatchOperationFailedException batchException = null;
+
+    for ( GenericFilePath path : paths ) {
+      try {
+        copyFile( path, destinationPath );
+      } catch ( OperationFailedException e ) {
+        if ( batchException == null ) {
+          batchException = new BatchOperationFailedException( "Error(s) occurred while attempting to copy files." );
+        }
+
+        batchException.addFailedPath( path, e );
+      }
+    }
+
+    if ( batchException != null ) {
+      throw batchException;
+    }
+  }
+
+  @Override
+  public void copyFile( @NonNull GenericFilePath path, @NonNull GenericFilePath destinationPath )
+    throws OperationFailedException {
+    getOwnerFileProvider( path ).copyFile( path, destinationPath );
```

Comment:
Not sure where to validate this, if here or within the provider: that `destinationPath` is also owned  by the same provider that `path` belongs to.

We're not supporting inter-provider moves or copies, atm, but, in the future, it would have to be performed by this layer. This feature might also force us to have more granular methods for providers, in that they would e.g. receive an InputStream or return an OutputStream, and possibly others.